File: ruby/lib/boardly/features/post.rb
# frozen_string_literal: true

require_relative "../notify/slack"

module Boardly
  module Features
    # Post a report to GitHub (a comment on a fixed issue or a fresh issue) and
    # then broadcast it to any configured external channels (Slack/email).
    module Post
      module_function

      def report(ctx, target, feature, fallback_title, body)
        owner = ctx.run_repo[:owner]
        repo = ctx.run_repo[:repo]

        if !target[:issue].nil?
          ctx.audit.record(feature, "comment", "#{owner}/#{repo}##{target[:issue]}", fallback_title)
          ctx.client.comment(owner, repo, target[:issue], body) unless ctx.dry_run
        else
          title = target[:create_issue_title] || fallback_title
          ctx.audit.record(feature, "create-issue", "#{owner}/#{repo}", title)
          ctx.client.create_issue(owner, repo, title, body, target[:labels]) unless ctx.dry_run
        end

        ctx.notifier.broadcast(Notify::Report.new(feature: feature, title: fallback_title, markdown: body))
      end
    end
  end
end

File: src/features/post.ts
import type { RunContext } from "./context.js";

export interface PostTarget {
  issue?: number;
  createIssueTitle?: string;
  labels: string[];
}

/** Post a report body either as a comment on a fixed issue or as a fresh issue. */
export async function postReport(
  ctx: RunContext,
  target: PostTarget,
  feature: string,
  fallbackTitle: string,
  body: string,
): Promise<void> {
  const { runRepo, client, audit } = ctx;

  if (target.issue !== undefined) {
    audit.record(feature, "comment", `${runRepo.owner}/${runRepo.repo}#${target.issue}`, fallbackTitle);
    if (!ctx.dryRun) {
      await client.comment(runRepo.owner, runRepo.repo, target.issue, body);
    }
    return;
  }

  const title = target.createIssueTitle ?? fallbackTitle;
  audit.record(feature, "create-issue", `${runRepo.owner}/${runRepo.repo}`, title);
  if (!ctx.dryRun) {
    await client.createIssue(runRepo.owner, runRepo.repo, title, body, target.labels);
  }
}
